### PR TITLE
nvidia_smi: Not count users with zero memory allocated

### PR DESCRIPTION
**Pull request by scatenag**
Exclude from nr_user count users with zero memory allocated

##### Summary
Really small change;
just to avoid counting users with actually no memory allocated.

##### Component Name
nvidia-smi python collector

### CODE REVIEW

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6361660&bid=20647858)

**Comment by ilyam8** [comment]
Hi

> just to avoid counting users with actually no memory allocated.

Why? The chart is `Memory Used by Each User in MiB`, zero is a value too. Why do you want to exclude zero memory users?

**Comment by ilyam8** [comment]
And there is gui option `Settings->Performance->Which dimensions to show?` to hide zero value dimensions.

**Comment by scatenag** [comment]
Well ... I was actually thinking if it made sense and how to make it have sense ...

Since GPU usage is not managed by nvidia-smi which consider only memory allocation,
in my setting, I use nr_user 
(that is the number of users with processes with memory allocated on a GPU)
for estimate GPU usage 
(and then for raising an alarm when two users use the same GPU and also for user resource accounting).

The fact is that some users allocate 0mb on all GPU without actually use them (they say by pytorch lightning), so I am using this patch.

I know it sounds very strange, but,
without the patch, I get a process with no memory

Screenshot_2020-10-19_12-41-59

and the user count goes to 1

image

**That is, I cannot consider a GPU used (by a user) with no memory allocated.**

"hide zero value dimensions" doesn't seem to mitigate the problem: nr_users is calculated from the collector - which consider also user with processes with 0 memory allocated - not from GUI.

It is not vital, but, for me, it is annoying to patch bach at every update...

I hope I have been able to explain...

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py [hunk 1]
```diff
@@ -404,7 +404,7 @@ def data(self):
         for p in processes:
             data['process_mem_{0}'.format(p['pid'])] = p['used_memory']
             if p['username']:
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:406]:
```python
if p['username'] and p['used_memory'] > 0:
```

**Comment by ilyam8** [comment]
> It is not vital, but, for me, it is annoying to patch bach at every update...

it is always possible to `cp nvidia_smi.chart.py my_nvidia_smi.chart.py`, make changes and disable `nvidia_smi` and enable `my_nvidia_smi` 😄

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py (hunk 1 above)
Comment (scatenag) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:406]: It is slightly different: in your way we not see the process (and the user) at all.
With my mod, I patch only user_nr ... 

I don't know which makes more sense...

**Comment by scatenag** [comment]
Yes, is what I do for a month ...
But I am not notified when netdata is updated and the process is quite boring having many machines with GPU ...

Definitely not vital: you can decide what is better for netdata...

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py (hunk 1 above)
Comment (ilyam8) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:406]: We would see the process.

This code that adds users
https://github.com/netdata/netdata/blob/0cb1d2a92e3aed369f32667128f65f68484267ca/collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py#L406-L412

You said 

> I cannot consider a GPU used (by a user) with no memory allocated.

So we need to skip those users, right?

Comment (scatenag) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:406]: we have to skip users for which used_memory is 0

`if p['used_memory'] > 0 : users.add(p['username'])`

**Comment by ilyam8** [comment]
> maybe this mess up long-term metrics ?

Yes.

> Definitely is not vital: you can decide what is better for netdata...

I think if `I cannot consider a GPU used (by a user) with no memory allocated` is true in general we can hardcode filtering 0 memory users, if it is your specific case then we need to make it optional (configurable: `exclude_zero_memory_allocated_users: yes` or something).

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6361660&bid=20647858)

File: collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py (hunk 1 above)
Comment (ilyam8) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:406]: it is not becase `key = 'user_mem_{0}'.format(p['username'])`, what am i missing?

Comment (ilyam8) [reply collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py:406]: process memory usage is `p['used_memory']`

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6361660&bid=20647858)

**Comment by scatenag** [comment]
Yes, optional (configurable: exclude_zero_memory_allocated_users: yes) is better;
I tried to realize it but I broke it on  line 460, in get_data
`self.exclude_zero_memory_allocated_users = configuration.get('exclude_zero_memory_allocated_users', False).`

Could you try to fix it?
Thanks.

**Comment by ilyam8** [comment]
> Could you try to fix it?

Yes, give me a minute!

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6361660&bid=20647858)

**Comment by ilyam8** [comment]
@scatenag i belive i've fixed the issue, git pull your branch and test it!

**Comment by ilyam8** [comment]
> Exclude from nr_user count users with zero memory allocated

Hold on. So you want to exclude those users from `Number of users` chart but have them in `User memory` chart? Or just to exclude them in general is ok?

**Comment by scatenag** [comment]
If we left on processes maybe it is fine to leave also on users mem ...
In my case ignore them just in user_nr is fine; then looking at user_mem and process mem one can see more in detail...  
You can decide.

**Comment by ilyam8** [comment]
@scatenag ok test current implementaion and if it works for you then let it be like that!

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6361660&bid=20647858)

**Comment by scatenag** [comment]
Everything works as expected.
Thanks a lot!!

File: collectors/python.d.plugin/nvidia_smi/README.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ Sample:
 ```yaml
 loop_mode    : yes
 poll_seconds : 1
+exclude_zero_memory_allocated_users : yes
```
Comment (ilyam8) [inline collectors/python.d.plugin/nvidia_smi/README.md:55]:
```suggestion
exclude_zero_memory_users : yes
```

**Review by ilyam8 — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
LGTM!

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- collectors/python.d.plugin/nvidia_smi/nvidia_smi.chart.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=6361660&bid=20647858)

**Pull request merged**